logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/JoinSubconversationUseCase.kt

```diff
@@ -0,0 +1,101 @@
+package com.wire.kalium.logic.feature.conversation
+
+import com.wire.kalium.logic.CoreFailure
+import com.wire.kalium.logic.NetworkFailure
+import com.wire.kalium.logic.data.conversation.MLSConversationRepository
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.id.GroupID
+import com.wire.kalium.logic.data.id.toApi
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.flatMap
+import com.wire.kalium.logic.functional.flatMapLeft
+import com.wire.kalium.logic.kaliumLogger
+import com.wire.kalium.logic.wrapApiRequest
+import com.wire.kalium.network.api.base.authenticated.conversation.ConversationApi
+import com.wire.kalium.network.api.base.authenticated.conversation.SubconversationDeleteRequest
+import com.wire.kalium.network.exceptions.KaliumException
+import com.wire.kalium.network.exceptions.isMlsStaleMessage
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import kotlinx.datetime.toInstant
+import kotlin.time.Duration
+
+/**
+ * Join a sub-conversation of a MLS conversation
+ */
+interface JoinSubconversationUseCase {
+    suspend operator fun invoke(conversationId: ConversationId, subconversationId: String): Either<CoreFailure, Unit>
+}
+
+class JoinSubconversationUseCaseImpl(
+    val conversationApi: ConversationApi,
+    val mlsConversationRepository: MLSConversationRepository
+) : JoinSubconversationUseCase {
+    override suspend operator fun invoke(conversationId: ConversationId, subconversationId: String): Either<CoreFailure, Unit> =
+        joinOrEstablishSubconversationAndRetry(conversationId, subconversationId)
+    private suspend fun joinOrEstablishSubconversation(
+        conversationId: ConversationId,
+        subconversationId: String
+    ): Either<CoreFailure, Unit> =
+        wrapApiRequest {
+            conversationApi.fetchSubconversationDetails(conversationId.toApi(), subconversationId)
+        }.flatMap { subconversationDetails ->
+            if (subconversationDetails.epoch > 0UL) {
+                val durationSinceLastEpoch = (subconversationDetails.epochTimestamp?.toInstant()?.timeElapsedUntilNow()?.inWholeHours ?: 0)
```

Comment:
maybe an extension function to do that?

Reply:
On which Class? I already added one on `Instant`. But I agree that line is very long :) Maybe just unwrapping the Instant?